docs/ckeditor/plugin.md

```diff
@@ -0,0 +1,476 @@
+
+Anyhow, that's all we need to do for the conversion. A lot more documentation on this topic can be found [here](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/deep-dive/conversion/custom-element-conversion.html). Overall, it is not terribly difficult to add a custom type in this way, and as long as we keep our HTML representations different and unambiguous we shouldn't run into any problems.
+
+### Putting the editing plugin together
+
+After we've written `_defineSchema` and `_defineConverters` we're pretty much done, all we need to do is make sure that those methods are both called in the `init` method.
+
+Here's the full source for this plugin:
+
+```ts
+class YoutubeEmbedEditing extends Plugin {
+  init() {
+    this._defineSchema()
+    this._defineConverters()
+  }
+
+  _defineSchema() {
+    const schema = this.editor.model.schema
+
+    schema.register("youtubeEmbed", {
+      isObject:       true,
+      allowWhere:     "$block",
+      allowContentOf: "$block"
+    })
+  }
+
+  _defineConverters() {
+    const conversion = this.editor.conversion
+
+    /**
+     * `upcast` converts the HTML string to a view element (i.e. ckeditor
+     * internal state, *not* to a DOM element)
+     *
+     * providing 'iframe' and YOUTUBE_EMBED_CLASS allows CKEditor to recognize
+     * the iframe in our converted HTML and convert it to a `youtubeEmbed` object
+     */
+    conversion.for("upcast").elementToElement({
+      model: "youtubeEmbed",
+      view:  {
+        name:    "section",
+        classes: YOUTUBE_EMBED_CLASS
+      }
+    })
+
+    /**
+     * dataDowncast converts a view element to an HTML element for data
+     * output (this output will then be the input into our html2md function)
+     */
+    conversion.for("dataDowncast").elementToElement({
+      model: "youtubeEmbed",
+      view:  {
+        name:    "section",
+        classes: YOUTUBE_EMBED_CLASS
+      }
+    })
+
+    /**
+     * editingDowncast converts a view element to HTML which is actually shown
+     * in the editor for WYSIWYG purposes
+     * (for the youtube embed this is an iframe)
+     */
+    conversion.for("editingDowncast").elementToElement({
+      model: "youtubeEmbed",
+      view:  (modelElement: any, { writer: viewWriter }: any) => {
+        // this looks bad but I promise it's fine
+        const videoId = modelElement._children._nodes[0]._data
+        const iframe = viewWriter.createContainerElement("iframe", {
+          class: YOUTUBE_EMBED_CLASS,
+          src:   youtubeEmbedUrl(videoId),
+          ...YOUTUBE_EMBED_PARAMS
+        })
+        return toWidget(iframe, viewWriter, { label: "Youtube Embed" })
+      }
+    })
+  }
+}
+```
```

Comment:
This looks like a great example to me, especially with all of the comments. I think this section could be briefer and more straightforward if it was rewritten like this:

```
### Define converters

(same intro up to the iframe example)

(full code example above)

(Notes explaining upcast/downcast etc. in more detail, 
links to relevant CKEditor docs, `toWidget`, anything else 
that isn't covered in code comments) 
```

To me it seems like that format could get rid of some redundant details, but you can take or leave that suggestion.

Reply:
I like having a separate section for each key part, and then a 'putting it together' section which is consistent across the things, so I'm inclined to leave it as-is